tchannel/container/heap.py

```diff
@@ -0,0 +1,86 @@
+from __future__ import absolute_import
+
+
+class HeapOperation(object):
+    def cmp(self, i, j):
+        """Compare the items in position i and j of the heap.
+
+        :param i: the first item's position of the heap list
+        :param j: the second item's position of the heap list
+        :return true if items[i] < items[j], otherwise false.
+        """
+        raise NotImplementedError()
+
+    def push(self, x):
+        """Push a new item into heap"""
+        raise NotImplementedError()
+
+    def pop(self):
+        """Pop an item from the heap"""
+        raise NotImplementedError()
+
+    def swap(self, i, j):
+        """swap items between position i and j of the heap"""
+        raise NotImplementedError()
+
+
+def init(h):
+    # heapify
+    n = len(h)
+    for i in range(n/2 - 1, -1):
+        down(h, i, n)
+
+
+def push(h, x):
+    h.push(x)
+    up(h, len(h)-1)
```

Comment:
Given that the HeapOperation has custom swap/compare/etc., it should also have a custom length operation so that you don't have to implement `__len__`.